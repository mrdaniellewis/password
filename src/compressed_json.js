export function stringify(data) {
  let count = 0;
  const keys = {};
  const compacted = JSON.stringify(data, (key, value) => {
    if (value && typeof value === 'object' && !Array.isArray(value)) {
      const replaced = {};
      Object.keys(value).forEach((k) => {
        if (!keys[k]) {
          keys[k] = `$${count}`;
          count += 1;
        }
        replaced[keys[k]] = value[k];
      });
      return replaced;
    }
    return value;
  });
  return JSON.stringify([
    Object.keys(keys),
    JSON.parse(compacted),
  ]);
}

export function parse(data) {
  const [keys, compacted] = JSON.parse(data);
  return JSON.parse(JSON.stringify(compacted), (key, value) => {
    if (value && typeof value === 'object' && !Array.isArray(value)) {
      const replaced = {};
      Object.keys(value).forEach((k) => {
        replaced[keys[parseInt(k.slice(1), 10)]] = value[k];
      });
      return replaced;
    }
    return value;
  });
}
