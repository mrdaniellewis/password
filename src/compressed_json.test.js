import { stringify, parse } from './compressed_json';

const testArray = [
  null,
  true,
  false,
  0,
  'string',
  [1, 2],
  { '': 'foo', foe: 'thumb' },
];

const testObject = {
  fee: 'fi',
  array: testArray,
  object: { '': 'foo', foe: 'thumb' },
};

describe('stringify', () => {
  it('stringifies an array', () => {
    const stringified = stringify(testArray);
    expect(stringified).toMatchSnapshot();
  });

  it('stringifies an object', () => {
    const stringified = stringify(testObject);
    expect(stringified).toMatchSnapshot();
  });
});

describe('parse', () => {
  it('parses an array', () => {
    const stringified = stringify(testArray);
    expect(parse(stringified)).toEqual(testArray);
  });

  it('parses an object', () => {
    const stringified = stringify(testObject);
    expect(parse(stringified)).toEqual(testObject);
  });
});
