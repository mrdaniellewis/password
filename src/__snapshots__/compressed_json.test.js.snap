// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringify stringifies an array 1`] = `"[[\\"\\",\\"foe\\"],[null,true,false,0,\\"string\\",[1,2],{\\"$0\\":\\"foo\\",\\"$1\\":\\"thumb\\"}]]"`;

exports[`stringify stringifies an object 1`] = `"[[\\"fee\\",\\"array\\",\\"object\\",\\"\\",\\"foe\\"],{\\"$0\\":\\"fi\\",\\"$1\\":[null,true,false,0,\\"string\\",[1,2],{\\"$3\\":\\"foo\\",\\"$4\\":\\"thumb\\"}],\\"$2\\":{\\"$3\\":\\"foo\\",\\"$4\\":\\"thumb\\"}}]"`;
